### PR TITLE
feat(map_loader): add publish map projector info

### PR DESCRIPTION
**Pull request by tkhmy**
## Description
- add publish map projector info in the map loader
- https://github.com/tier4/tier4_autoware_msgs/pull/80
- separated from https://github.com/autowarefoundation/autoware.universe/pull/2930
- This info is used to reverse the pose to gps point

### related PR
https://github.com/autowarefoundation/autoware_common/pull/174
https://github.com/tier4/tier4_autoware_msgs/pull/80
https://github.com/autowarefoundation/autoware.universe/pull/3492

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by isamu-takagi** [comment]
@tkhmy 
When I used `local_x` and `local_y` it would print the wrong code. I think it should output an empty code in this case.
```xml
<node id='20001' visible='true' version='1' lat='0' lon='0'>
  <tag k='local_x' v='3752.410888671875' />
  <tag k='local_y' v='73740.9609375' />
  <tag k='ele' v='19.31' />
</node>
```

```bash
$ ros2 topic echo /map/mgrs_grid
data: 31NAA
```

**Comment by tkhmy** [comment]
> @tkhmy When I used `local_x` and `local_y` it would output the wrong code. I think it should output an empty code in this case.
> 
> ```
> <node id='20001' visible='true' version='1' lat='0' lon='0'>
>   <tag k='local_x' v='3752.410888671875' />
>   <tag k='local_y' v='73740.9609375' />
>   <tag k='ele' v='19.31' />
> </node>
> ```
> 
> ```shell
> $ ros2 topic echo /map/mgrs_grid
> data: 31NAA
> ```

@isamu-takagi I think I fix in this comit 1002366, can you have a try?

**Comment by isamu-takagi** [comment]
> I think I fix in this comit 1002366, can you have a try?

@tkhmy MGRS code is always empty.

**Comment by tkhmy** [comment]
> Shouldn't we also think about supporting UTM projection as well?

We going to support UTM projection as well.

**Comment by tkhmy** [comment]
@isamu-takagi @mitsudome-r  @KYabuuchi 
Hi, sorry for the delay, I added the publish for UTM and also local projector.
Can you guys review?

File: map/map_loader/config/lanelet2_map_loader.param.yaml
Comment (isamu-takagi) [inline map/map_loader/config/lanelet2_map_loader.param.yaml:3]: Local is auto-detected so it is not used here.

File: map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp
Comment (isamu-takagi) [inline map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp:117]: Here we only detect local attributes. If it is detected, override lanelet2_map_projector_type to local.

Comment (tkhmy) [reply map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp:117]: But I think most of the map will have local attributes if is generated by VMB.
If we only detect local attributes, most of the map will be local

Comment (isamu-takagi) [reply map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp:117]: If the map has a local attribute, it cannot be MGRS or UTM. Because there is no guarantee that the local coordinates will match the latitude and longitude. Therefore maps containing local attributes cannot support global pose. I think the VMB should not output redundant local attributes.

File: map/map_loader/config/lanelet2_map_loader.param.yaml
Comment (isamu-takagi) [reply map/map_loader/config/lanelet2_map_loader.param.yaml:3]: We decided to use local in https://github.com/autowarefoundation/autoware/issues/3427.

File: map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp
Comment (isamu-takagi) [reply map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp:117]: We decided to use local type in https://github.com/autowarefoundation/autoware/issues/3427.

Comment (mitsudome-r) [inline map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp:117]: I think this should be created as a different PR since adding "local" option to the loader and adding a feature to publish map projector info is different.

Comment (tkhmy) [reply map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp]: @mitsudome-r 
The code is reverted and move to https://github.com/autowarefoundation/autoware.universe/pull/3492

**Comment by isamu-takagi** [comment]
Wait for merging https://github.com/autowarefoundation/autoware.universe/pull/3492 to add code for local type.

**Comment by tkhmy** [comment]
@isamu-takagi
I added local support, please review, thanks

**Review by isamu-takagi — APPROVED** [review]

**Comment by tkhmy** [comment]
@mitsudome-r @KYabuuchi 
Hi, can you guys review as the code owner?
The message is in this https://github.com/tier4/tier4_autoware_msgs/pull/80

**Review by KYabuuchi — APPROVED** [review]
I have verified this and it seems to work properly. :ok_man:

**Comment by kminoda** [comment]
@tkhmy What is the status of this PR? (waiting for reviews?)

**Comment by tkhmy** [comment]
@kminoda I want to merge it, but looks like the merge permission is still not enough.
I think because I edited this file, @takayuki5168 @kosuke55 , can you help me to review and approve it if it looks fine for you?
`planning/static_centerline_optimizer/src/static_centerline_optimizer_node.cpp`

**Review by takayuki5168 — APPROVED** [review]

**Comment by tkhmy** [comment]
@kminoda Looks like you also one of the maintainer instead of Yabuuchi san.
Can you help me to approve it?

File: map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp
Comment (kminoda) [inline map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp:83]: (correct me if I am wrong)
Since this function does not use member variables, how about making it independent function from `Lanelet2MapLoaderNode` class?

File: map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp
Comment (kminoda) [inline map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp:129]: This function as well

**Review by kminoda — COMMENTED** [review]
Sure thing, I added some comments :pray:

**Comment by tkhmy** [comment]
@kminoda I added some change, can you have a look?

**Review by kminoda — APPROVED** [review]
LGTM

**Pull request merged**
